fix(isNumericString): better error handling

yumetodo opened the pull request:
To detect weather string is numeric correctly, we need to check the message from strtod by following below:

1. no conversion performed: return value is 0, the value of nptr(1st argument) is stored in the object pointed to by endptr(2nd argument)
2. other error: check errno value

However, it is impossible via strtod to check weather input string cause underflows portably
because whether errno acquires the value ERANGE is implementation-defined.

ref:
- https://qiita.com/yumetodo/items/238751b879c09b56234b

yumetodo committed:
fix(isNumericString): better error handling

To detect weather string is numeric correctly, we need to check the message from strtod by following below:

1. no conversion performed: return value is 0, the value of nptr(1st argument) is stored in the object pointed to by endptr(2nd argument)
2. other error: check errno value

However, it is impossible via strtod to check weather input string cause underflows portably
because whether errno acquires the value ERANGE is implementation-defined.

ref:
- https://qiita.com/yumetodo/items/238751b879c09b56234b

yumetodo committed:
fix(isNumericString): fix condition

s/||/&&/

test: https://wandbox.org/permlink/tWMem36riCNiOKyu

yumetodo commented:
> (Here I expect `isNumericString(s)` returns true if and only if `s` is completely numeric; for example, `-20`, `12.4` and `3e+05`, not `32.4AWF`, `hoge23` and `1 2 3`. Also I expect `isNumericString("1e99999") == true`, `isNumericString("1e-99999") == true`. Scratch 3.0 seems to cast a string to a number like these.)

I'm trying to fix this issue, please wait for a moment.

yumetodo committed:
fix(isNumericString): ignore when errno is ERANGE

yumetodo commented:
@yos1up fixed!

test: https://wandbox.org/permlink/VmifOtJH1gqvWv1F

yos1up commented:
Thanks for fixing. I think that the condition `!(0 == re && s.c_str() == ep) && 0 == errno && NULL != ep && '\0' == ep[0]` is a bit too long, and this condition can be reduced to the equivalent condition `s[0] != '\0' && NULL != ep && '\0' == ep[0]`, because of the following reasons:

- The value of `errno` can be either `0` or `ERANGE` just after `strtod(...)`, and the latter is immediately substituted by `0` in the next line, so the condition `errno == 0` is always true.

- (Former ⇒ Latter) If `!(0 == re && s.c_str() == ep)` is true, `s != ""` is also true, and therefore `s[0] != '\0'` is also true.

- (Latter ⇒ Former) Inversely, if the conditions `s[0] != '\0'` and  `'\0' == ep[0]` is true, `s.c_str() == ep` is false, then `!(0 == re && s.c_str() == ep)` is true.

Then, the return value can be reduced to `s[0] != '\0' && NULL != ep && '\0' == ep[0]` (or more shortly `s[0] && ep && !*ep`). 

Anyway, in the current master code, the return value is `!ep || !*ep`, which bahaves wrongly for `s == ""` case (it wrongly returns true). The new condition discussed above nicely returns false in such case.

yumetodo committed:
fix(isNumericString): simplify condition

Now we can ignore errno in condition!
However, ignoring errno marked the beginning of making side-effects.

yumetodo commented:
@yos1up Thank you for your review! fixed!

yos1up commented:
Thank you!

yumetodo commented:
>In Scratch '000' is regarded as non-numeric

to fix this issue, in C++11 or later, include `algorithm` header and change like below:

```patch
-        strtod(s.c_str(), &ep);
+        const double re = strtod(s.c_str(), &ep);
        //Scratch 3.0 recognize the string cause underflows or overflows as Numeric
-        return s[0] != '\0' && NULL != ep && '\0' == ep[0];
+        return s[0] != '\0' && NULL != ep && '\0' == ep[0] && !(0 == re && std::all_of(s.begin(), s.end(), [](char c){ return c == '0'; }));
```

ref: https://cpprefjp.github.io/reference/algorithm/all_of.html

(BTW, this project can use C++11 feature?)
